### PR TITLE
Adding ACTION and CREATE endpoints

### PR DESCRIPTION
**Pull request by Imsiaocong**

### CODE REVIEW

File: .vagrant/rgloader/loader.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# This file loads the proper rgloader/loader.rb file that comes packaged
+# with Vagrant so that encoded files can properly run with Vagrant.
+
+if ENV["VAGRANT_INSTALLER_EMBEDDED_DIR"]
```
Comment (cslecrone) [inline .vagrant/rgloader/loader.rb:4]: We don't want to commit anything under the `.vagrant` directory, and it should be in a `.gitignore` file. Speaking of @avllee21, how's that coming along? 

@Imsiaocong, can you please remove this file from your pull request?

File: service/models.py [hunk 2]
```diff
@@ -54,9 +54,19 @@ class InventoryItem(db.Model):
...
+    in_stock = db.Column(db.Boolean(), nullable=False, default=False) # updated June 29
 
     def __repr__(self):
         return f"<Inventory item {self.sku} id={self.id}>"
+        
+    def update(self):
+        """
+        Updates an Inventory item to the database
+        """
+        logger.info("Saving %s",self.sku) # TODO: Not sure about the sku part. Might change it later.
```
Comment (cslecrone) [inline service/models.py:66]: There should be a space between your comma and `self.sku`.

File: service/models.py [hunk 3]
```diff
@@ -67,6 +77,7 @@ def serialize(self) -> Dict[str, Union[str, int]]:
             "condition": self.condition.name,
             "restock_level": self.restock_level,
             "restock_amount": self.restock_amount,
+            "in_stock": self.in_stock, # updated June 29
```
Comment (cslecrone) [inline service/models.py:80]: No need for the `#updated June 29th` comment, it will be in our `git log` :D

File: service/models.py [hunk 4]
```diff
@@ -77,6 +88,7 @@ def deserialize(self, data: Dict[str, Union[str, int]]):
             self.condition = getattr(Condition, data["condition"])
             self.restock_level = data["restock_level"]
             self.restock_amount = data["restock_amount"]
+            self.in_stock = data["in_stock"] # updated June 29
```
Comment (cslecrone) [inline service/models.py:91]: Same note here, the comma is redundant when we have version control.

File: service/models.py [hunk 5]
```diff
@@ -97,3 +109,17 @@ def init_db(cls, app: Flask):
         app.app_context().push()
         # Create tables
         db.create_all()
+
+    @classmethod
+    def find(cls, inventory_item_id):
+        """Finds a Pet by it's ID
```
Comment (cslecrone) [inline service/models.py:115]: Looks like this is copy-paste and still has references to `Pet`. Can you clean this up?

Also, is this for the `ACTION` endpoint, or one of the `QUERY` endpoints?

File: service/status.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# coding: utf8
+# Copyright 2016, 2021 John J. Rofrano. All Rights Reserved.
```
Comment (cslecrone) [inline service/status.py:2]: We probably want our own copyright header and not John's :)

File: unittests.xml [hunk 7]
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8"?><testsuite name="nosetests" tests="5" errors="0" failures="0" skip="0"><testcase classname="tests.test_inventory_items.TestInventoryItemModel" name="test_deserialize" time="0.028"></testcase><testcase classname="tests.test_inventory_items.TestInventoryItemModel" name="test_deserialize_bad_data" time="0.019"></testcase><testcase classname="tests.test_inventory_items.TestInventoryItemModel" name="test_deserialize_missing_data" time="0.016"></testcase><testcase classname="tests.test_inventory_items.TestInventoryItemModel" name="test_repr" time="0.016"></testcase><testcase classname="tests.test_inventory_items.TestInventoryItemModel" name="test_serialize" time="0.059"></testcase></testsuite>
```
Comment (cslecrone) [inline unittests.xml:1]: Same thing here, we don't want to commit `unittests.xml`. @Imsiaocong, can you please remove this file from the Pull Request?

**Review by cslecrone — CHANGES_REQUESTED** [review]
Cool, thanks for the PR!

There are a handful of issues here. First off, `.vagrant/rgloader/loader.rb` and `unittests.xml` shouldn't be checked in to version control. We'll want to remove them. 

Next, you're missing unit tests for the new code that you're adding, and it appears that your changes broke a couple of the existing tests:

```
(venv) vagrant@ubuntu:/vagrant$ nosetests

Test Cases for InventoryItem Model
- Test deserialization of an InventoryItem (ERROR)
- Test deserialization of an InventoryItem with bad data
- Test deserialization of an InventoryItem with missing data
- Ensure string representation of an InventoryItem is correct (ERROR)
- Test serialization of an InventoryItem

======================================================================
ERROR: Test deserialization of an InventoryItem
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/tests/test_inventory_items.py", line 105, in test_deserialize
    item.deserialize(data)
  File "/vagrant/service/models.py", line 93, in deserialize
    raise DataValidationError(
service.models.DataValidationError: Invalid inventory item: missing in_stock
-------------------- >> begin captured logging << --------------------
flask.app: INFO: Initializing database
faker.factory: DEBUG: Not in REPL -> leaving logger event level as is.
flask.app: INFO: Initializing database
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Ensure string representation of an InventoryItem is correct
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/service/models.py", line 91, in deserialize
    self.in_stock = data["in_stock"] # updated June 29
KeyError: 'in_stock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/vagrant/tests/test_inventory_items.py", line 160, in test_repr
    item.deserialize(
  File "/vagrant/service/models.py", line 93, in deserialize
    raise DataValidationError(
service.models.DataValidationError: Invalid inventory item: missing in_stock

----------------------------------------------------------------------
XML: /vagrant/unittests.xml
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
service/__init__.py      26      4    85%   36, 46-49
service/models.py        58      8    86%   60, 66-69, 100, 124-125
---------------------------------------------------
TOTAL                    84     12    86%
----------------------------------------------------------------------
Ran 5 tests in 0.319s

FAILED (errors=2)
```
Remember that we need to stay at 95% coverage. There should be tests you can refer to in `lab-flask-tdd` to help guide you here.

File: service/status.py (hunk 6 above)
Comment (Imsiaocong) [reply service/status.py:2]: lol didn't notice that

**Review by Imsiaocong — COMMENTED** [review]
> Cool, thanks for the PR!
> 
> There are a handful of issues here. First off, `.vagrant/rgloader/loader.rb` and `unittests.xml` shouldn't be checked in to version control. We'll want to remove them.
> 
> Next, you're missing unit tests for the new code that you're adding, and it appears that your changes broke a couple of the existing tests:
> 
> ```
> (venv) vagrant@ubuntu:/vagrant$ nosetests
> 
> Test Cases for InventoryItem Model
> - Test deserialization of an InventoryItem (ERROR)
> - Test deserialization of an InventoryItem with bad data
> - Test deserialization of an InventoryItem with missing data
> - Ensure string representation of an InventoryItem is correct (ERROR)
> - Test serialization of an InventoryItem
> 
> ======================================================================
> ERROR: Test deserialization of an InventoryItem
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/vagrant/tests/test_inventory_items.py", line 105, in test_deserialize
>     item.deserialize(data)
>   File "/vagrant/service/models.py", line 93, in deserialize
>     raise DataValidationError(
> service.models.DataValidationError: Invalid inventory item: missing in_stock
> -------------------- >> begin captured logging << --------------------
> flask.app: INFO: Initializing database
> faker.factory: DEBUG: Not in REPL -> leaving logger event level as is.
> flask.app: INFO: Initializing database
> --------------------- >> end captured logging << ---------------------
> 
> ======================================================================
> ERROR: Ensure string representation of an InventoryItem is correct
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/vagrant/service/models.py", line 91, in deserialize
>     self.in_stock = data["in_stock"] # updated June 29
> KeyError: 'in_stock'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/vagrant/tests/test_inventory_items.py", line 160, in test_repr
>     item.deserialize(
>   File "/vagrant/service/models.py", line 93, in deserialize
>     raise DataValidationError(
> service.models.DataValidationError: Invalid inventory item: missing in_stock
> 
> ----------------------------------------------------------------------
> XML: /vagrant/unittests.xml
> Name                  Stmts   Miss  Cover   Missing
> ---------------------------------------------------
> service/__init__.py      26      4    85%   36, 46-49
> service/models.py        58      8    86%   60, 66-69, 100, 124-125
> ---------------------------------------------------
> TOTAL                    84     12    86%
> ----------------------------------------------------------------------
> Ran 5 tests in 0.319s
> 
> FAILED (errors=2)
> ```
> 
> Remember that we need to stay at 95% coverage. There should be tests you can refer to in `lab-flask-tdd` to help guide you here.

Thanks for reviewing Cam. It appears the error comes from the new field "in_stock" I added to our InventoryItem model. Working on it.

File: tests/test_inventory_items.py [hunk 8]
```diff
@@ -89,6 +94,17 @@ def test_serialize(self):
         self.assertIn("restock_amount", data)
         self.assertEqual(data["restock_amount"], item.restock_amount)
 
+        # Same as above, but for in_stock
+        self.assertIn("in_stock", data)
+        self.assertEqual(data["in_stock"], item.in_stock)
+
+    def test_create_a_inventory_item(self):
+        """Create an inventory item and assert that it exists"""
+        inventory_item = InventoryItem(sku="fido", count=10, condition="New", restock_level=2, restock_amount=20, in_stock=True)
```
Comment (cslecrone) [inline tests/test_inventory_items.py:103]: This line is a bit long, the standard is 80 characters for PEP8.

**Review by cslecrone — CHANGES_REQUESTED** [review]
Hey @Imsiaocong, this is really close!

Just a couple of last things to address. 

First off, there's still a `.DS-Store` in your PR, which needs to be removed. 

Second, it looks like there's some PyLint stuff that needs to be cleaned up as well:
```
(venv) vagrant@ubuntu:/vagrant$ pylint service/ | grep -v __init__.py
************* Module service
************* Module service.models
service/models.py:1:2: W0511: TODO: Add a license header (fixme)
service/models.py:75:2: W0511: TODO: Not sure about the sku part. Might change it later. (fixme)
service/models.py:51:27: C0303: Trailing whitespace (trailing-whitespace)
service/models.py:52:23: C0303: Trailing whitespace (trailing-whitespace)
service/models.py:61:0: C0303: Trailing whitespace (trailing-whitespace)
service/models.py:75:0: C0301: Line too long (102/100) (line-too-long)
service/models.py:134:0: C0304: Final newline missing (missing-final-newline)
service/models.py:17:0: C0103: Constant name "logger" doesn't conform to UPPER_CASE naming style (invalid-name)
service/models.py:20:0: C0103: Constant name "db" doesn't conform to UPPER_CASE naming style (invalid-name)
service/models.py:23:12: W0621: Redefining name 'app' from outer scope (line 14) (redefined-outer-name)
service/models.py:31:4: W0107: Unnecessary pass statement (unnecessary-pass)
service/models.py:67:8: C0103: Attribute name "id" doesn't conform to snake_case naming style (invalid-name)
service/models.py:47:9: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:47:19: E1101: Instance of 'SQLAlchemy' has no 'Integer' member (no-member)
service/models.py:48:10: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:48:20: E1101: Instance of 'SQLAlchemy' has no 'String' member (no-member)
service/models.py:49:12: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:49:22: E1101: Instance of 'SQLAlchemy' has no 'Integer' member (no-member)
service/models.py:50:16: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:51:8: E1101: Instance of 'SQLAlchemy' has no 'Enum' member (no-member)
service/models.py:55:20: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:55:30: E1101: Instance of 'SQLAlchemy' has no 'Integer' member (no-member)
service/models.py:56:21: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:56:31: E1101: Instance of 'SQLAlchemy' has no 'Integer' member (no-member)
service/models.py:57:15: E1101: Instance of 'SQLAlchemy' has no 'Column' member (no-member)
service/models.py:57:25: E1101: Instance of 'SQLAlchemy' has no 'Boolean' member (no-member)
service/models.py:68:8: E1101: Instance of 'scoped_session' has no 'add' member (no-member)
service/models.py:69:8: E1101: Instance of 'scoped_session' has no 'commit' member (no-member)
service/models.py:78:8: E1101: Instance of 'scoped_session' has no 'commit' member (no-member)
service/models.py:112:21: W0621: Redefining name 'app' from outer scope (line 14) (redefined-outer-name)
service/models.py:14:0: W0611: Unused app imported from flask (unused-import)
************* Module service.routes
service/routes.py:1:2: W0511: TODO: Add a license header (fixme)
service/routes.py:31:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:35:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:36:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:37:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:38:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:39:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:40:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:41:0: C0301: Line too long (101/100) (line-too-long)
service/routes.py:41:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:43:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:44:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:45:0: C0330: Wrong hanging indentation (remove 4 spaces).
                jsonify(message), status.HTTP_201_CREATED, {"Location": location_url}
            |   ^ (bad-continuation)
service/routes.py:53:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:58:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:59:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:60:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:61:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:62:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:63:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:64:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:65:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:67:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:68:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:79:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:80:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:81:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:82:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:83:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:84:0: W0312: Found indentation with tabs instead of spaces (mixed-indentation)
service/routes.py:85:0: C0330: Wrong hanging indentation (remove 4 spaces).
                status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            |   ^ (bad-continuation)
service/routes.py:86:0: C0330: Wrong hanging indentation (remove 4 spaces).
                "Content-Type must be {}".format(media_type),
            |   ^ (bad-continuation)
service/routes.py:87:0: C0304: Final newline missing (missing-final-newline)
service/routes.py:35:1: E1101: Method 'logger' has no 'info' member (no-member)
service/routes.py:43:1: E1101: Method 'logger' has no 'info' member (no-member)
service/routes.py:58:1: E1101: Method 'logger' has no 'info' member (no-member)
service/routes.py:67:1: E1101: Method 'logger' has no 'info' member (no-member)
service/routes.py:74:0: W0105: String statement has no effect (pointless-string-statement)
service/routes.py:83:1: E1101: Method 'logger' has no 'error' member (no-member)
service/routes.py:11:0: W0611: Unused import os (unused-import)
service/routes.py:12:0: W0611: Unused import sys (unused-import)
service/routes.py:13:0: W0611: Unused import logging (unused-import)
service/routes.py:14:0: W0611: Unused Flask imported from flask (unused-import)
service/routes.py:20:0: W0611: Unused SQLAlchemy imported from flask_sqlalchemy (unused-import)
service/routes.py:21:0: W0611: Unused DataValidationError imported from service.models (unused-import)
service/routes.py:16:0: C0411: third party import "from werkzeug.exceptions import NotFound" should be placed before "from . import status" (wrong-import-order)
service/routes.py:20:0: C0411: third party import "from flask_sqlalchemy import SQLAlchemy" should be placed before "from . import status" (wrong-import-order)
service/routes.py:21:0: C0411: first party import "from service.models import InventoryItem, DataValidationError" should be placed before "from . import status" (wrong-import-order)
************* Module service.status
service/status.py:1:0: R0401: Cyclic import (service -> service.routes) (cyclic-import)

--------------------------------------------------------------------
Your code has been rated at -2.69/10 (previous run: -2.69/10, +0.00)
```
Happy to approve once those two things are addressed.

File: service/models.py [hunk 9]
```diff
@@ -54,9 +54,19 @@ class InventoryItem(db.Model):
     )
     restock_level = db.Column(db.Integer, nullable=False)
     restock_amount = db.Column(db.Integer, nullable=False)
+    in_stock = db.Column(db.Boolean(), nullable=False, default=False) # updated June 29
```
[the remark this answers is not in the document]
Comment (Imsiaocong) [reply service/models.py]: Hi, can you please elaborate on this?

[the remark this answers is not in the document]
Comment (cslecrone) [reply service/models.py]: You already did this! Should be good to go.

[the remark this answers is not in the document]
Comment (Imsiaocong) [reply service/models.py]: Oh that's fine. I didn't really notice the outdated label

[the remark this answers is not in the document]
Comment (Imsiaocong) [reply service/models.py]: Just pushed the modified.

**Pull request merged**